source/Port.cpp

```diff
@@ -0,0 +1,203 @@
+#include "GameData.h"
+#include "Government.h"
+#include "PlayerInfo.h"
+#include "Politics.h"
+#include "Random.h"
+#include "Ship.h"
+#include "ShipEvent.h"
+#include "SpriteSet.h"
+#include "System.h"
+
+#include <algorithm>
+
+using namespace std;
+
+namespace {
+	const string SPACEPORT = "Space_port";
+}
+
+
+
+// Load a port's description from a node.
+void Port::Load(const DataNode &node)
+{
+	loaded = true;
+	if(node.Size() > 1)
+		name = node.Token(1);
+
+	for(const DataNode &child : node)
+	{
+		const string &key = child.Token(0);
+
+		if(key == "recharges" && (child.HasChildren() || child.Size() >= 2))
+		{
+			auto setRecharge = [&](const DataNode &valueNode, const string &value) noexcept -> void {
+				if(value == "all")
+					recharge |= RechargeType::All;
+				else if(value == "shields")
+					recharge |= RechargeType::Shields;
+				else if(value == "hull")
+					recharge |= RechargeType::Hull;
+				else if(value == "energy")
+					recharge |= RechargeType::Energy;
+				else if(value == "fuel")
+					recharge |= RechargeType::Fuel;
+				else
+					valueNode.PrintTrace("Skipping unrecognized attribute:");
+			};
+			for(int i = 1; i < child.Size(); ++i)
+				setRecharge(child, child.Token(i));
+			for(const DataNode &grand : child)
+				setRecharge(grand, grand.Token(0));
+		}
+		else if(key == "services" && (child.HasChildren() || child.Size() >= 2))
+		{
+			auto setServices = [&](const DataNode &valueNode, const string &value) noexcept -> void {
+				if(value == "all")
+					services |= ServicesType::All;
+				else if(value == "trading")
+					services |= ServicesType::Trading;
+				else if(value == "job board")
+					services |= ServicesType::JobBoard;
+				else if(value == "bank")
+					services |= ServicesType::Bank;
+				else if(value == "hire crew")
+					services |= ServicesType::HireCrew;
+				else if(value == "offers missions")
+					services |= ServicesType::OffersMissions;
+				else
+					valueNode.PrintTrace("Skipping unrecognized attribute:");
+			};
+			for(int i = 1; i < child.Size(); ++i)
+				setServices(child, child.Token(i));
+			for(const DataNode &grand : child)
+				setServices(grand, grand.Token(0));
+		}
+		else if(key == "news")
+			hasNews = true;
+		else if(key == "description" && child.Size() >= 2)
+		{
+			const string &value = child.Token(1);
+			if(!description.empty() && !value.empty() && value[0] > ' ')
+				description += '\t';
+			description += value;
+			description += '\n';
+
+			// If we have a description but no name then use the default spaceport name.
+			if(name.empty())
+				name = SPACEPORT;
+		}
+		else
+			child.PrintTrace("Skipping unrecognized attribute:");
+	}
+}
+
+
+
+void Port::LoadDefaultSpaceport()
+{
+	name = SPACEPORT;
+	recharge = RechargeType::All;
+	services = ServicesType::All;
+	hasNews = true;
+}
+
+
+
+void Port::LoadUninhabitedSpaceport()
+{
+	name = SPACEPORT;
+	recharge = RechargeType::All;
+	services = ServicesType::OffersMissions;
+	hasNews = false;
```

Comment:
This matches the old behavior.
```suggestion
	hasNews = true;
```